eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jpa/LegacyJpaEventStorageEngine.java

```diff
@@ -0,0 +1,625 @@
+import org.axonframework.eventhandling.TrackedEventData;
+import org.axonframework.eventhandling.TrackedEventMessage;
+import org.axonframework.eventhandling.TrackingToken;
+import org.axonframework.eventsourcing.eventstore.AggregateBasedConsistencyMarker;
+import org.axonframework.eventsourcing.eventstore.AppendCondition;
+import org.axonframework.eventsourcing.eventstore.AsyncEventStorageEngine;
+import org.axonframework.eventsourcing.eventstore.ConsistencyMarker;
+import org.axonframework.eventsourcing.eventstore.EventCriteria;
+import org.axonframework.eventsourcing.eventstore.LegacyAggregateBasedEventStorageEngineUtils;
+import org.axonframework.eventsourcing.eventstore.LegacyResources;
+import org.axonframework.eventsourcing.eventstore.SourcingCondition;
+import org.axonframework.eventsourcing.eventstore.StreamingCondition;
+import org.axonframework.eventsourcing.eventstore.TaggedEventMessage;
+import org.axonframework.messaging.Context;
+import org.axonframework.messaging.MessageStream;
+import org.axonframework.messaging.MessageType;
+import org.axonframework.messaging.MetaData;
+import org.axonframework.serialization.Serializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Spliterators;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static org.axonframework.common.BuilderUtils.*;
+import static org.axonframework.common.ObjectUtils.getOrDefault;
+import static org.axonframework.eventsourcing.eventstore.LegacyAggregateBasedEventStorageEngineUtils.*;
+import static org.axonframework.eventsourcing.eventstore.LegacyAggregateBasedEventStorageEngineUtils.resolveAggregateIdentifier;
+
+
+/**
+ * EventStorageEngine implementation that uses JPA to store and fetch events.
+ * <p>
+ * By default, the payload of events is stored as a serialized blob of bytes. Other columns are used to store meta-data
+ * that allow quick finding of DomainEvents for a specific aggregate in the correct order.
+ *
+ * @author Mateusz Nowak
+ * @since 5.0.0
+ */
+public class LegacyJpaEventStorageEngine implements AsyncEventStorageEngine {
+
+    private static final Logger logger = LoggerFactory.getLogger(LegacyJpaEventStorageEngine.class);
+    private static final String DOMAIN_EVENT_ENTRY_ENTITY_NAME = DomainEventEntry.class.getSimpleName();
+
+    private final EntityManagerProvider entityManagerProvider;
+    private final TransactionManager transactionManager;
+    private final Serializer eventSerializer;
```

Comment:
I propose merging this PR in its current state, if it is good enough, and addressing the remaining work (e.g., changing the mechanism from serializer to converter, deciding how to handle metadata serialization, etc.) in subsequent PRs. This approach keeps the PR manageable in size and allows others to use the updated test suite and incorporate the bug fixes.

Reply:
I agree, once my comments have been processed ;-)
If some of those turn out to large to tackle, we can discuss on a per-comment basis if we should move those to another PR.